cpp/src/stream_compaction/distinct.cu

```diff
@@ -80,21 +81,62 @@ std::unique_ptr<table> distinct(table_view const& input,
   // insert distinct indices into the map.
   key_map.insert(iter, iter + num_rows, hash_key, key_equal, stream.value());
 
+#if 0
   auto const output_size{key_map.get_size()};
   auto distinct_indices = cudf::make_numeric_column(
     data_type{type_id::INT32}, output_size, mask_state::UNALLOCATED, stream, mr);
   // write distinct indices to a numeric column
   key_map.retrieve_all(distinct_indices->mutable_view().begin<cudf::size_type>(),
                        thrust::make_discard_iterator(),
                        stream.value());
-
   // run gather operation to establish new order
   return detail::gather(input,
                         distinct_indices->view(),
                         out_of_bounds_policy::DONT_CHECK,
                         detail::negative_index_policy::NOT_ALLOWED,
                         stream,
                         mr);
+#else
+
+  auto const key_size{keys_view.num_rows()};
+  auto distinct_indices = rmm::device_uvector<size_type>(key_size, stream);
+
+  // Fill `key_size` if keep_first
+  // Fill `INT_MIN` if keep_last
+  thrust::uninitialized_fill(rmm::exec_policy(stream),
+                             distinct_indices.begin(),
+                             distinct_indices.begin() + key_size,
+                             key_size);
+
+  auto const d_map = key_map.get_device_view();
+  thrust::for_each(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(key_size),
+    [distinct_indices = distinct_indices.begin(), d_map, hash_key, key_equal] __device__(auto key) {
+      // iter should always be valid, because all keys have been inserted.
+      // Here idx is the index of the unique elements that has been inserted into the map.
+      // As such, `find` calling for all duplicate keys will return the same idx value.
+      auto const idx =
+        d_map.find(key, hash_key, key_equal)->second.load(cuda::std::memory_order_relaxed);
+
+      // Store the minimum index of all keys that are equal.
+      atomicMin(&distinct_indices[idx], key);
```

Comment:
The idea looks great!

I never thought about doing atomic operations during retrieval. I was mainly thinking about how to use `cuco::static_reduction_map` to "keep first/last" during insertion.